src/core/utils/mesh_diag.cpp

```diff
@@ -0,0 +1,286 @@
+        }
+    }
+
+    mDiscoverCallback.Set(aCallback, aContext);
+    mTimer.Start(kResponseTimeout);
+
+exit:
+    return error;
+}
+
+Error MeshDiag::SendDiagGetTo(uint16_t aRloc16, const DiscoverConfig &aConfig)
+{
+    Error            error   = kErrorNone;
+    Coap::Message   *message = nullptr;
+    Tmf::MessageInfo messageInfo(GetInstance());
+    uint8_t          tlvs[kMaxTlvsToRequest];
+    uint8_t          tlvsLength;
+
+    message = Get<Tmf::Agent>().NewConfirmablePostMessage(kUriDiagnosticGetRequest);
+    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+
+    IgnoreError(message->SetPriority(Message::kPriorityLow));
+
+    tlvs[0]    = Address16Tlv::kType;
+    tlvs[1]    = ExtMacAddressTlv::kType;
+    tlvs[2]    = RouteTlv::kType;
+    tlvsLength = 3;
+
+    if (aConfig.mDiscoverIp6Addresses)
+    {
+        tlvs[tlvsLength++] = Ip6AddressListTlv::kType;
+    }
+
+    if (aConfig.mDiscoverChildTbale)
+    {
+        tlvs[tlvsLength++] = ChildTableTlv::kType;
+    }
+
+    SuccessOrExit(error = Tlv::Append<TypeListTlv>(*message, tlvs, tlvsLength));
+
+    messageInfo.SetSockAddrToRlocPeerAddrTo(aRloc16);
+    error = Get<Tmf::Agent>().SendMessage(*message, messageInfo, HandleDiagGetResponse, this);
+
+exit:
+    FreeMessageOnError(message, error);
+    return error;
+}
+
+void MeshDiag::HandleDiagGetResponse(void                *aContext,
+                                     otMessage           *aMessage,
+                                     const otMessageInfo *aMessageInfo,
+                                     Error                aResult)
+{
+    static_cast<MeshDiag *>(aContext)->HandleDiagGetResponse(AsCoapMessagePtr(aMessage), AsCoreTypePtr(aMessageInfo),
+                                                             aResult);
+}
+
+void MeshDiag::HandleDiagGetResponse(Coap::Message *aMessage, const Ip6::MessageInfo *aMessageInfo, Error aResult)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    Error           error;
+    RouterInfo      routerInfo;
+    Ip6AddrIterator ip6AddrIterator;
+    ChildIterator   childIterator;
+
+    SuccessOrExit(aResult);
+    VerifyOrExit((aMessage != nullptr) && mTimer.IsRunning());
+
+    SuccessOrExit(routerInfo.ParseFrom(*aMessage));
+
+    if (ip6AddrIterator.InitFrom(*aMessage) == kErrorNone)
+    {
+        routerInfo.mIp6AddrIterator = &ip6AddrIterator;
+    }
+
+    if (childIterator.InitFrom(*aMessage, routerInfo.mRloc16) == kErrorNone)
+    {
+        routerInfo.mChildIterator = &childIterator;
+    }
+
+    mExpectedRouterIdSet.Remove(routerInfo.mRouterId);
+
+    if (mExpectedRouterIdSet.GetNumberOfAllocatedIds() == 0)
+    {
+        error = kErrorNone;
+        mTimer.Stop();
+    }
+    else
+    {
+        error = kErrorPending;
+    }
+
+    mDiscoverCallback.InvokeIfSet(error, &routerInfo);
+
+exit:
+    return;
+}
+
+void MeshDiag::HandleTimer(void)
+{
+    // Timed out waiting for response from one or more routers.
```

Comment:
Should we rely on the CoAP timeout instead? We can configure the CoAP parameters to something other than the default if needed.

Reply:
Yes. We can use the CoAP response timeout and I was about to change this but thinking more on this, I expect we will need a timer for other features/commands we want to add to `MeshDiag`, so may be easier to use the `mTimer` for this as well.

Reply:
Relying on the CoAP timeout helps ensure that the CoAP transaction has ended.

Alternatively, we could use a separate timeout but should also explicitly cancel the CoAP transaction in the timeout handler.

Reply:
Good point on canceling them. 
I added `Get<Tmf::Agent>().AbortTransacation()` from `Cancel()` but not from timer handler. Will add it.

Reply:
Given a 5 second timeout and the default CoAP parameters (2 second initial retransmission timeout), there will only be one CoAP retransmission.

Reply:
Yes. I expect so.  Sounds ok to me to allow one retx before overall timeout? What do you think?

- Default coap parm is `ack-timeout = 2 s` with `factor = 3/2 = 1.5`
- So initial wait is randomly selected in `[2, 3]` seconds.
- After than we double the retx/wait time 
  - So even if `2` is picked, first retx after `2`, second one would be at `2+2*2=6` which won't happen if we use overall timeout of `5` sec..

We can define different CoAP params for this, but not sure if we can have better values:
- One thing that I sort of not like about CoAP retx model, is the initial random interval that is picked will be reused for all next attempts (same value is doubled). 
- So a device picking shorter initial time will retx faster and finish earlier (unless we use `factor` very close to `1` which does not sound good).

Reply:
I think this is fine for now. We can always adjust later as we get more operational experience.

Reply:
@abtink This is a well-tested pattern: retry with exponential backoff. The increasing distance between retries of different devices is a feature of this pattern and not a bug. (https://en.wikipedia.org/wiki/Exponential_backoff)